src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -33,55 +31,84 @@ internal abstract partial class AbstractNavigateToSearchService
         private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
             Project project, Document searchDocument, string pattern, CancellationToken cancellationToken)
         {
-            var containsDots = pattern.IndexOf('.') >= 0;
-            using (var patternMatcher = new PatternMatcher(pattern, allowFuzzyMatching: true))
+            // If the user created a dotted pattern then we'll grab the last part of the name
+            var (patternName, patternContainerOpt) = PatternMatcher.GetNameAndContainer(pattern);
+            var nameMatcher = PatternMatcher.CreatePatternMatcher(patternName, includeMatchedSpans: true, allowFuzzyMatching: true);
+
+            var containerMatcher = patternContainerOpt != null
+                ? PatternMatcher.CreateDotSeperatedContainerMatcher(patternContainerOpt)
+                : null;
+
+            using (nameMatcher)
+            using (containerMatcher)
             {
-                var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
-                foreach (var document in project.Documents)
+                var nameMatches = ArrayBuilder<PatternMatch>.GetInstance();
+                var containerMatches = ArrayBuilder<PatternMatch>.GetInstance();
+
+                try
                 {
-                    if (searchDocument != null && document != searchDocument)
-                    {
-                        continue;
-                    }
+                    return await FindNavigableDeclaredSymbolInfosAsync(
+                        project, searchDocument, nameMatcher, containerMatcher,
+                        nameMatches, containerMatches, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    nameMatches.Free();
+                    containerMatches.Free();
+                }
+            }
+        }
+
+        private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
+            Project project, Document searchDocument,
+            PatternMatcher nameMatcher, PatternMatcher containerMatcher,
+            ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches,
+            CancellationToken cancellationToken)
+        {
+            var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+            foreach (var document in project.Documents)
+            {
+                if (searchDocument != null && document != searchDocument)
+                {
+                    continue;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
:memo: This call to `ThrowIfCancellationRequested` appears unnecessary. I know you like following a pattern of putting it near the beginning of methods and loops, but in my experience a great cancellation experience can be maintained with fewer calls by targeting locations prior to "performing work" (e.g. the call within the `foreach` below). Assuming that `GetSyntaxTreeIndexAsync` performs reasonably with respect to cancellation requests is not a bad assumption. This leads to cleaner, and in aggregate slightly better performing, code.